File: backtracking/rec_subsets/rec_subsets.cpp
// string::substr
#include <iostream>
#include <string>

void RecSubsets(std::string soFar, std::string rest){
    if (rest.empty()) {
        std::cout << soFar << std::endl;
    } else {
        //include the first char
        RecSubsets(soFar + rest[0], rest.substr(1));
        // exclude the first char
        RecSubsets(soFar, rest.substr(1));
    }
}


int main ()
{
    std::string a1 = "Good";
    std::string a2 = "";
    RecSubsets(a2, a1);
    return 0;
}

File: backtracking/rec_subsets/rec_subsets.py
"""rec_subsets.py

use the python3 to run this code
python3 rec_subsets.py
"""


def rec_subsets(so_far, rest):
    """This method finds the possible combinations for given string

    :param so_far: possible combinations of string.
    :type so_far: str
    :param rest: specify the actual string.
    :type rest: str
    :returns: None
    :raises: None
    """
    if not rest:
        print(so_far)
    else:
        # include the first char
        rec_subsets(so_far + rest[0], rest[1:])
        # exclude the first char
        rec_subsets(so_far, rest[1:])


def main():
    """Main function
    """
    a = 'Good'
    rec_subsets("", a)


if __name__ == '__main__':
    main()
